src/ejabberd_c2s.erl

```diff
@@ -2682,16 +2682,17 @@ fsm_limit_opts(Opts) ->
     end.
 
 
--spec bounce_messages() -> 'ok'.
-bounce_messages() ->
+-spec bounce_messages(StateData :: state()) -> 'ok'.
+bounce_messages(State) ->
     receive
         {route, From, To, El} ->
+            notify_unacknowledged_msg(State, {From, To, El}),
```

Comment:
Won't they be handled by a new session or by ordinary event pusher logic (if the user is offline)?

Reply:
:+1:

Reply:
> by the ordinary pusher logic - no, there's no subscription for offline message hook

@DenysGonchar I don't understand why - `route/3` should trigger the `user_send_packet` hook which in turns is processed by the event pusher hook translator. Could you explain?

Reply:
hm. MB you're right. actually this PR is just as-is copy of Safaricom changes. they use OMEMO encryption (so message can be decrypted only once, after that decryption keys are removed) and has unique ID for each message. so i believe they simply ignore the second notification.

actually my intention was to merge only hook code, agree that pusher notifications must be rechecked twice and reworked.

Reply:
but hook code itself must be covered with some tests as well